src/plugins/discover/public/plugin.ts

```diff
@@ -99,24 +152,171 @@ export class DiscoverPlugin implements Plugin<DiscoverSetup, DiscoverStart> {
       component: JsonCodeBlock,
     });
 
+    const {
+      appMounted,
+      appUnMounted,
+      stop: stopUrlTracker,
+      setActiveUrl: setTrackedUrl,
+    } = createKbnUrlTracker({
+      // we pass getter here instead of plain `history`,
+      // so history is lazily created (when app is mounted)
+      // this prevents redundant `#` when not in discover app
+      getHistory,
+      baseUrl: core.http.basePath.prepend('/app/kibana'),
+      defaultSubUrl: '#/discover',
+      storageKey: `lastUrl:${core.http.basePath.get()}:discover`,
+      navLinkUpdater$: this.appStateUpdater,
+      toastNotifications: core.notifications.toasts,
+      stateParams: [
+        {
+          kbnUrlKey: '_g',
+          stateUpdate$: plugins.data.query.state$.pipe(
+            filter(
+              ({ changes }) => !!(changes.globalFilters || changes.time || changes.refreshInterval)
+            ),
+            map(({ state }) => ({
+              ...state,
+              filters: state.filters?.filter(esFilters.isFilterPinned),
+            }))
+          ),
+        },
+      ],
+    });
+    setUrlTracker({ setTrackedUrl });
+    this.stopUrlTracking = () => {
+      stopUrlTracker();
+    };
+
+    this.docViewsRegistry.setAngularInjectorGetter(this.getEmbeddableInjector);
+    plugins.kibanaLegacy.registerLegacyApp({
+      id: 'discover',
+      title: 'Discover',
+      updater$: this.appStateUpdater.asObservable(),
+      navLinkId: 'kibana:discover',
+      order: -1004,
+      euiIconType: 'discoverApp',
+      mount: async (params: AppMountParameters) => {
+        if (!this.initializeServices) {
+          throw Error('Discover plugin method initializeServices is undefined');
+        }
+        if (!this.initializeInnerAngular) {
+          throw Error('Discover plugin method initializeInnerAngular is undefined');
+        }
+        appMounted();
+        const {
+          plugins: { data: dataStart },
+        } = await this.initializeServices();
+        await this.initializeInnerAngular();
+
+        // make sure the index pattern list is up to date
+        await dataStart.indexPatterns.clearCache();
+        const { renderApp } = await import('./application/application');
+        const unmount = await renderApp(innerAngularName, params.element);
```

Comment:
I know you are already doing some ping-pong  between `setup` and `start` for this `initializeInnerAngular` call, but ideally, `initializeInnerAngular ` would be included in the async import instead just the app (`await import('./application/application')`)

That would avoid 

```
import { getInnerAngularModuleEmbeddable, getInnerAngularModule } from './get_inner_angular';
```

to be present in the root plugin file. Seing all the things this `get_inner_angular` module imports, that would probably drastically reduce the `discover` plugin initial bundle's size.

I see that `getInnerAngularModuleEmbeddable` is used outside of the `mount` function, so this change seems really non-trivial, but that would really be a nice performances improvement.

That could be done in a follow-up PR though.

Reply:
@pgayvallet  This is a really good point!
But if you don't mind, I would do it in a follow up PR in scope of https://github.com/elastic/kibana/issues/63596